File: 21553 - Sai Nandan D/Week 8/34_Find First and Last Position of Element in Sorted Array.py
# Approach 1: Recursive
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums: return [-1, -1]
        def search(lo, hi):
            if nums[lo] == target == nums[hi]:
                return [lo, hi]
            if nums[lo] <= target <= nums[hi]:
                mid = (lo + hi) // 2
                l, r = search(lo, mid), search(mid+1, hi)
                return max(l,r) if -1 in l+r else [l[0], r[1]]
            return [-1, -1]
        return search(0, len(nums)-1)
        

# Approach 2.1: Iterative
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def search(target):
            lo, hi = 0, len(nums)
            while lo < hi:
                mid = (lo + hi) // 2
                if nums[mid] < target:
                    lo = mid + 1
                else:
                    hi = mid
            return lo
                    
        lo = search(target)
        return [lo, search(target+1)-1] if target in nums[lo:lo+1] else [-1, -1]
        
# Approach 2.2: Iterative (Using library functions)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:                    
        lo = bisect.bisect_left(nums, target)
        return [lo, bisect.bisect(nums, target)-1] if target in nums[lo:lo+1] else [-1, -1]

File: 21561 - Suvam Tamang/Week 8/34_Find First and Last Position of Element in Sorted Array.java
class Solution 
{
    public int[] searchRange(int[] nums, int target) 
    {
        int low = 0, high = nums.length-1, mid, res1 = -1;
        while(low <= high)
        {
            mid = (low + high) / 2;
            if(nums[mid] == target)
            {
                res1 = mid;
                high = mid - 1; 
            }
            else if(nums[mid] < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        
        low = 0;
        high = nums.length - 1;
        int res2 = -1;
        
        while(low <= high)
        {
            mid = (low + high) / 2;
            if(nums[mid] == target)
            {
                res2 = mid;
                low = mid + 1; 
            }
            else if(nums[mid] < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        
        return new int[]{res1 ,res2};
    }
}
